Add Jolokia agent in proxy mode

whale-shark opened the pull request:
Modified the object of sending to support Jolokia proxy mode.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

ruflin commented:
@whale-shark Thanks a lot for opening this PR. Could you make Houndci happy and add an entry to the CHANGELOG.asciidoc file? I'm pretty sure this change will make quite a few people happy.

I was wondering if there is a good way to have automated tests for this?

whale-shark commented:
@ruflin I plan to carry out CI correspondence and document addition on the weekend.

I've just started Go language, and I'm also studying about testing.
Jolokia's test did not work for both `make test-module [no test files]` and `go test jmx_integration_test.go` from the beginning.
I also want to do a good test and good automated test.

Absolutely test is necessary?

ruflin commented:
I wonder if jolokia can be run in proxy mode also if it's run on the same machine? Best would be to have also a docker container with the full setup we then can connect to to test this the same way as we do it now. Perhaps we could even have 2 instances of Jolokia running, one in proxy mode and one not?

The main reason I'm asking for tests is because in case it breaks I'm looking for an easy way to have a setup to also test it manually. And there is obviously the benefit of the automated way of knowing when it breaks.

Please post any questions you have for the setup or beats here, more then happy to help you get going.

ruflin commented:
@whale-shark Any update here? Would be great to move this forward.

whale-shark commented:
Sorry for late reply.
I'm busy with my job and I've not been able enough time to create a test.

I am considering adding an instance of docker to connect to remote JMX.

Please wait just a little longer.

whale-shark commented:
I am sorry to be late.
Unit test and integration test added.

I changed Dockerfile.
In the original configuration, it can not be launched in proxy mode.
So I deployed jolokia's war to Tomcat to work in proxy mode.
Also, we added a new Tomcat to be connected in proxy mode.

ruflin commented:
@whale-shark Any chance you could rebase this on the most recent version of master?

jsoriano commented on CHANGELOG.asciidoc:
```diff
@@ -519,6 +519,13 @@ https://github.com/elastic/beats/compare/v6.0.1...v6.1.0[View commits]
+- Update the MySQL dashboard to use the Time Series Visual Builder. {pull}5996[5996]
+- Add experimental uwsgi module. {pull}6006[6006]
+- Docker and Kubernetes modules are now GA, instead of Beta. {pull}6105[6105]
+- Add Jolokia agent in proxy mode. {pull}6475[6475]
```
@whale-shark something weird happened with the changelog after updating the branch, could you remove the added lines that are not related with your change?

jsoriano commented:
jenkins, test it please

jsoriano commented:
jenkins, test this please

whale-shark commented:
@jsoriano Thank you for pointing out

I revert to latest branch and added my fix line.

jsoriano commented:
jenkins, test it please

jsoriano commented:
jenkins, test this

jsoriano commented on metricbeat/module/jolokia/_meta/Dockerfile:
```diff
@@ -1,11 +1,39 @@
+
+# Set up a server to be connected in proxy mode
+RUN wget http://archive.apache.org/dist/tomcat/tomcat-9/v${PROXT_TARGET_TOMCAT_VERSION}/bin/${PX_TC}.tar.gz;\
+    tar xzf ${PX_TC}.tar.gz -C /usr;\
```
Can you also remove these .tar.gz files that are not needed in the final image?

whale-shark replied:
I deleted the unnecessary tar.gz file from the final image.

jsoriano commented on metricbeat/module/jolokia/_meta/Dockerfile:
```diff
@@ -1,11 +1,39 @@
+    -Dcom.sun.management.jmxremote.ssl=false\
+    -Dcom.sun.management.jmxremote.authenticate=true"\
+    /usr/${PX_TC}/bin/catalina.sh run & \
+    env CATALINA_OPTS="-Dorg.jolokia.jsr160ProxyEnabled" /usr/${TC}/bin/catalina.sh run
```
Would it be possible to run everything in an only tomcat?

whale-shark replied:
In order to check the connection in Proxy mode, Tomcat to be connected is added.
If you use in an only Tomcat, you will not know if you are connected to another Tomcat in the process of connecting in Proxy mode of the integration test.

jsoriano replied:
Ok, makes sense, maybe we could think in a two container scenario, but I think we can continue as is.

whale-shark replied:
Changed the connecting to another Tomcat to a pattern where jmx password authentication cannot be performed correctly.
Fix to be possible to run everything in an only Tomcat.

jsoriano commented:
jenkins, test this please

jsoriano commented:
jenkins, test this please

whale-shark commented:
@jsoriano  Fixed a file conflicting with the latest source.

jsoriano commented:
jenkins, test this again, please

jsoriano commented:
Failing tests are not related.
